File: app/backend/models/repair.py
"""Pydantic models for the repair and verification API endpoints."""

from typing import Optional
from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Repair request / change models
# ---------------------------------------------------------------------------

class RepairRequest(BaseModel):
    """Request body for POST /api/repair/preview and POST /api/repair/apply."""

    repository_path: str
    finding_ids: Optional[list[str]] = None
    """Optional list of check_id values to target.  If None, all findings are targeted."""
    dry_run: bool = False


class RepairChange(BaseModel):
    """A single proposed or applied file modification."""

    check_id: str
    file_path: str
    line: Optional[int]
    description: str
    old_text: str
    new_text: str
    status: str
    """One of: PROPOSED, AUTO_REPAIRED, MANUAL_REVIEW, SKIPPED, FAILED."""
    safe_to_apply: bool


class RepairResponse(BaseModel):
    """Response body for POST /api/repair/preview and POST /api/repair/apply."""

    repository_path: str
    dry_run: bool
    changes: list[RepairChange]
    applied_count: int
    skipped_count: int
    failed_count: int
    manual_review_count: int
    backup_created: bool
    backup_path: Optional[str]
    message: str


# ---------------------------------------------------------------------------
# Verification models
# ---------------------------------------------------------------------------

class VerificationFindingDetail(BaseModel):
    """Per-finding verification result."""

    check_id: str
    file_path: str
    line: Optional[int]
    title: str
    resolved: bool


class VerificationResponse(BaseModel):
    """Response body for POST /api/repair/verify."""

    repository_path: str
    tests_passed: int
    tests_failed: int
    analyzer_before_count: int
    analyzer_after_count: int
    resolved_count: int
    remaining_count: int
    regression_detected: bool
    verification_status: str
    """One of: PASSED, FAILED, PARTIAL, NOT_RUN."""
    details: list[VerificationFindingDetail]
    test_output: str


# ---------------------------------------------------------------------------
# Repair-session state (stored in memory for the current server session)
# ---------------------------------------------------------------------------

class RepairSession(BaseModel):
    """In-memory record of a repair session — one per repository path."""

    repository_path: str
    original_findings: list[dict]
    proposed_changes: list[dict]
    applied_changes: list[dict]
    manual_findings: list[str]
    backup_path: Optional[str]
    test_result: Optional[str]
    before_finding_count: int
    after_finding_count: Optional[int]
    resolved_finding_ids: list[str]
    remaining_finding_ids: list[str]

File: app/frontend/src/api/repair.ts
/** API types and functions for the repair and verification endpoints. */

const BASE = 'http://127.0.0.1:8000/api'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface RepairRequest {
  repository_path: string
  finding_ids?: string[]
  dry_run?: boolean
}

export interface RepairChange {
  check_id: string
  file_path: string
  line: number | null
  description: string
  old_text: string
  new_text: string
  /** PROPOSED | AUTO_REPAIRED | MANUAL_REVIEW | SKIPPED | FAILED */
  status: string
  safe_to_apply: boolean
}

export interface RepairResponse {
  repository_path: string
  dry_run: boolean
  changes: RepairChange[]
  applied_count: number
  skipped_count: number
  failed_count: number
  manual_review_count: number
  backup_created: boolean
  backup_path: string | null
  message: string
}

export interface VerificationFindingDetail {
  check_id: string
  file_path: string
  line: number | null
  title: string
  resolved: boolean
}

export interface VerificationResponse {
  repository_path: string
  tests_passed: number
  tests_failed: number
  analyzer_before_count: number
  analyzer_after_count: number
  resolved_count: number
  remaining_count: number
  regression_detected: boolean
  /** PASSED | FAILED | PARTIAL | NOT_RUN */
  verification_status: string
  details: VerificationFindingDetail[]
  test_output: string
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

/** Preview proposed repairs — repository is NOT modified. */
export async function previewRepairs(
  repositoryPath: string,
  findingIds?: string[],
): Promise<RepairResponse> {
  const response = await fetch(`${BASE}/repair/preview`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      repository_path: repositoryPath,
      finding_ids: findingIds,
      dry_run: true,
    } satisfies RepairRequest),
  })
  if (!response.ok) {
    const detail = await response.json().catch(() => ({ detail: response.statusText }))
    throw new Error(detail?.detail ?? `Preview failed with status ${response.status}`)
  }
  return response.json()
}

/** Apply safe automatic repairs to the repository. */
export async function applyRepairs(
  repositoryPath: string,
  findingIds?: string[],
): Promise<RepairResponse> {
  const response = await fetch(`${BASE}/repair/apply`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      repository_path: repositoryPath,
      finding_ids: findingIds,
      dry_run: false,
    } satisfies RepairRequest),
  })
  if (!response.ok) {
    const detail = await response.json().catch(() => ({ detail: response.statusText }))
    throw new Error(detail?.detail ?? `Apply failed with status ${response.status}`)
  }
  return response.json()
}

/** Run post-repair verification: pytest + re-analysis. */
export async function verifyRepairs(
  repositoryPath: string,
): Promise<VerificationResponse> {
  const response = await fetch(`${BASE}/repair/verify`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      repository_path: repositoryPath,
    } satisfies RepairRequest),
  })
  if (!response.ok) {
    const detail = await response.json().catch(() => ({ detail: response.statusText }))
    throw new Error(detail?.detail ?? `Verify failed with status ${response.status}`)
  }
  return response.json()
}
